File: Data Structures & Algorithms/subsets/submission-3.cpp
#include <vector>
class Solution {
private:
    std::vector<std::vector<int>> res;
public:
    vector<vector<int>> subsets(vector<int>& nums) {
        dfs({},0,nums);
        return res;
    }

    void dfs(std::vector<int> curr_lst,int i,std::vector<int> &nums){
        if (i==nums.size()) {
            res.push_back(curr_lst);
            return;
        }

        curr_lst.push_back(nums[i]);
        dfs(curr_lst,i+1,nums);
        curr_lst.pop_back();

        dfs(curr_lst,i+1,nums);
        return;
    }
};

File: Data Structures & Algorithms/climbing-stairs/submission-3.py
class Solution:
    def climbStairs(self, n: int) -> int:
        
        memo = {}
        def dfs(val):  
            if val in memo:
                return memo[val]
            if val == n:
                return 1
            elif val > n:
                return 0
            memo[val] = dfs(val+1) + dfs(val+2)
            return memo[val]
            
        return dfs(0)